### PR TITLE
add support for volume plugins

### PR DESCRIPTION
**Pull request by wlan0**
@ibuildthecloud The tests require docker1.9 dind to work.

### CODE REVIEW

File: cattle/plugins/docker/__init__.py [hunk 1]
```diff
@@ -40,7 +40,7 @@ def url_base():
 
     @staticmethod
     def api_version():
-        return default_value('DOCKER_API_VERSION', '1.18')
+        return default_value('DOCKER_API_VERSION', '1.21')
```
Comment (ibuildthecloud) [inline cattle/plugins/docker/__init__.py]: We need to keep this at 1.18, but just use 1.21 for volume stuff.

File: cattle/plugins/docker/storage.py [hunk 2]
```diff
@@ -175,10 +175,43 @@ def _get_volume_storage_pool_map_data(self, obj):
...
+        docker_client().create_volume(volume.name, driver, driverOpts)
+
     def _is_volume_inactive(self, volume, storage_pool):
-        return True
+        if not hasattr(volume, 'name'):
+            return True
+        if volume.name is None or volume.name == "":
+            return True
+        try:
+            docker_client().inspect_volume(volume.name)
```
Comment (ibuildthecloud) [inline cattle/plugins/docker/storage.py]: At this time can we pass the 1.21 version to create the client.  Create another config setting for storage_api_version, or something like that.

File: cattle/plugins/docker/storage.py [hunk 3]
```diff
@@ -175,10 +175,43 @@ def _get_volume_storage_pool_map_data(self, obj):
         }
 
     def _is_volume_active(self, volume, storage_pool):
+        if hasattr(volume.data, 'fields'):
```
Comment (ibuildthecloud) [inline cattle/plugins/docker/storage.py]: You've got to be able to simplify this method.

File: tests/test_docker.py [hunk 4]
```diff
@@ -25,6 +25,20 @@ def test_volume_activate(agent, responses):
 
 
 @if_docker
+def test_volume_activate_driver(agent, responses):
```
Comment (ibuildthecloud) [inline tests/test_docker.py]: I know in CI this will fail, but can you get these tests passing locally?

Comment (ibuildthecloud) [inline tests/test_docker.py]: You should have a test for activate, deactivate, and remove.  Make sure the case where the volumes are not for a driver are covered.  They may be already, not totally sure.  Just look at the existing tests.

File: requirements.txt [hunk 5]
```diff
@@ -1,5 +1,5 @@
 mako
-git+https://github.com/rancherio/docker-py.git@1.2.3-rancher-3
+git+https://github.com/ibuildthecloud/docker-py.git@missing-api
```
Comment (ibuildthecloud) [inline requirements.txt]: We need to push this to rancher/docker-py.

File: tests/test_docker.py (hunk 4 above)
Comment (ibuildthecloud) [inline tests/test_docker.py]: Add a test where driverOpts are None and also set

File: cattle/plugins/docker/storage.py [hunk 6]
```diff
@@ -175,10 +176,45 @@ def _get_volume_storage_pool_map_data(self, obj):
...
 
+    def _do_volume_activate(self, volume, storage_pool, progress):
+        driver = ""
+        driverOpts = None
+        if hasattr(volume.data, 'fields'):
+            if hasattr(volume.data.fields, 'driver'):
+                driver = volume.data.fields['driver']
+            if hasattr(volume.data.fields, 'driverOpts'):
+                driverOpts = volume.data.fields['driverOpts']
+        v = default_value('DOCKER_API_VERSION', '1.21')
```
Comment (ibuildthecloud) [inline cattle/plugins/docker/storage.py]: docker_client(version=DockerConfig.volume_api_version())

**Comment by ibuildthecloud** [comment]
Cover:

activate
deactivate
remove

But consider we need to keep 1.18 API behavior for remove.
